src/buildercore/threadbare/operations.py

```diff
@@ -0,0 +1,638 @@
+def rcd(remote_working_dir):
+    "ensures all commands run are done from the given remote directory. if remote directory doesn't exist, command will not be run"
+    with state.settings(remote_working_dir=remote_working_dir):
+        yield
+
+
+@contextlib.contextmanager
+def hide(what=None):
+    "hides *all* output, regardless of `what` type of output is to be hidden."
+    with state.settings(quiet=True):
+        yield
+
+
+def _ssh_client(**kwargs):
+    """returns an instance of pssh.clients.native.SSHClient
+    if within a state context, looks for a client already in use and returns that if found.
+    if not found, creates a new one and stores it for later use."""
+
+    # parameters we're interested in and their default values
+    base_kwargs = {
+        # current user. sensible default but probably not what you want
+        "user": getpass.getuser(),
+        "host_string": None,
+        # TODO: parallel-ssh and fabric both look for the same ~4 possible keys and
+        # use the first one they find. implement that behaviour
+        "key_filename": os.path.expanduser("~/.ssh/id_rsa"),
+        "port": 22,
+    }
+    global_kwargs, user_kwargs, final_kwargs = handle(base_kwargs, kwargs)
+    final_kwargs["password"] = None  # always private keys
+    rename(final_kwargs, [("key_filename", "pkey"), ("host_string", "host")])
+
+    # if we're not using global state, return the new client as-is
+    env = state.ENV
+    if env.read_only:
+        return SSHClient(**final_kwargs)
+
+    client_map_key = "ssh_client"
+    client_key = subdict(final_kwargs, ["user", "host", "pkey", "port", "timeout"])
+    client_key = tuple(sorted(client_key.items()))
+
+    # otherwise, check to see if a previous client is available for this host
+    client_map = env.get(client_map_key, {})
+    if client_key in client_map:
+        return client_map[client_key]
+
+    # if not, create a new one and store it in the state
+
+    # https://parallel-ssh.readthedocs.io/en/latest/native_single.html#pssh.clients.native.single.SSHClient
+    client = SSHClient(**final_kwargs)
+
+    # disconnect session when leaving context manager
+    state.add_cleanup(lambda: client.disconnect())
+
+    client_map[client_key] = client
+    env[client_map_key] = client_map
+
+    return client
+
+
+def _execute(command, user, key_filename, host_string, port, use_pty, timeout):
+    """creates an SSHClient object and executes given `command` with the given parameters."""
+    client = _ssh_client(
+        user=user, host_string=host_string, key_filename=key_filename, port=port
+    )
+
+    shell = False  # handled ourselves
+    sudo = False  # handled ourselves
+    user = None  # user to sudo to
+    encoding = "utf-8"  # used everywhere
+
+    try:
+        # https://parallel-ssh.readthedocs.io/en/latest/native_single.html#pssh.clients.native.single.SSHClient.run_command
+        channel, host_string, stdout, stderr, stdin = client.run_command(
```

Comment:
that's a lot of arguments on the input and output side; any data structure like a dictionary where they could be stored together? Easier to give the whole concept a name e.g. a `dict` `commandOptions` (terrible name but you get the point) for all the other arguments than `command`

Reply:
`_execute` is the exception here. I made the function signature more verbose so there wouldn't be any ambiguity while writing tests. Outputs can't be helped and hiding them in a named tuple or fancy dict would obfuscate it. `_execute` shouldn't be used directly anyway. Threadbare *is* a library. I should look into generating some docs for it.

`operations` *does* suffer from the many-options problem in general though. Fabric defined it's `env` full of defaults that these operations would then pull values out of and then compare them with in an inconsistent adhoc way. The trade-off I've made in Threadbare is that these additional options are in a `**kwargs` bucket for each function. Unsupported args are removed and argument precedence is handled in a consistent way (see `handle`). The downside to this is you need to inspect the code to see if an argument is supported or not.